### PR TITLE
Test2.2

### PR DESCRIPTION
**Pull request by sudarson-25**
PR created for review

### CODE REVIEW

File: Training/Training/change.c [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+void OptimalChangeCalculator (int* coins, int rem, int* denominations) {
+   for (int i = 0; i < 4; i++) {
+      coins[i] = rem / denominations[i];
+      rem %= denominations[i];
+   }
+}
+
+static void Testcases (int* denominations) {
+   int input[] = { 95,75,67,79 }, expected[][4] = { 9,1,0,0,7,1,0,0,6,1,1,0,7,1,2,0 }, output[4];
```
Comment (Dakshita-Minocha) [inline Training/Training/change.c]:
```suggestion
   int input[] = { 95,75,67,79 }, expected[][4] = { { 9, 1, 0, 0 }, { 7, 1, 0, 0 }, { 6, 1, 1, 0 }, {7, 1, 2, 0 }, output[4];
```
2D arrays should be declared as such for readability.

File: Training/Training/change.c [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+   printf ("\nTestcases\n~~~~~~~~~\n");
+   for (int i = 0; i < sizeof (input) / sizeof (input[0]); i++) {
+      OptimalChangeCalculator (output, input[i], denominations);
+      int j;
+      for (j = 0; j < 4; j++)
```
Comment (Dakshita-Minocha) [inline Training/Training/change.c]:
```suggestion
      for (int j = 0; j < 4; j++)
```

File: Training/Training/change.c [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+         if (expected[i][j] != output[j]) {
+            printf ("Case %d: FAILED\n", i + 1);
+            break;
```
Comment (Dakshita-Minocha) [inline Training/Training/change.c]: Instead of breaking you can continue; Check for j is not required.

File: Training/Training/change.c [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+   int num;
+   do {
+      char numberStr[4096];
+      printf ("%s", prompt);
+      fgets (numberStr, sizeof (numberStr), stdin);
+      if (output = strchr (numberStr, '\n') != NULL) {
+         errno = 0;
+         char* endptr = NULL;
+         num = strtol (numberStr, &endptr, 10);
+         output = endptr != numberStr && *endptr == '\n' && errno != ERANGE;
```
Comment (Dakshita-Minocha) [inline Training/Training/change.c]:
```suggestion
         output = endptr != numberStr && *endptr == '\n' && errno != ERANGE || num >= 0;
```
The validity of the input should include the value of num. Also, this name does not indicate the purpose of the variable all.

File: Training/Training/change.c [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+      }
+      if (!output || num < 0) printf ("Invalid Input\n");
```
Comment (Dakshita-Minocha) [inline Training/Training/change.c]: A better solution would be to prompt the user to enter the correct amount with a description of what you are expecting in the prompt.

**Review by Dakshita-Minocha — CHANGES_REQUESTED** [review]

**Comment by Dakshita-Minocha** [comment]
Name PR Test2: _description_

File: Training/Training/change.c (hunk 3 above)
Comment (sudarson-25) [reply Training/Training/change.c]: If more than 1 expected value fails, the statement will get printed as many times.

**Review by Dakshita-Minocha — APPROVED** [review]

**Pull request closed without merging**